backend/src/store.ts

```diff
@@ -40,42 +62,219 @@ export class Store {
    */
-  static fileStore(dbPath: string, logger: BaseLogger): Store {
-    return new Store(DbClient.fileClient(dbPath, logger, SchemaPath));
+  static fileStore(
+    dbPath: string,
+    logger: BaseLogger,
+    machineId: string
+  ): Store {
+    return new Store(
+      DbClient.fileClient(dbPath, logger, SchemaPath),
+      machineId
+    );
   }
 
   /**
    * Builds and returns a new store whose data is kept in memory.
    */
-  static memoryStore(): Store {
-    return new Store(DbClient.memoryClient(SchemaPath));
+  static memoryStore(machineId: string): Store {
+    return new Store(DbClient.memoryClient(SchemaPath), machineId);
+  }
+
+  private applyEventsToVoters(voters: Voter[]): Voter[] {
+    const rows = this.client.all(
+      `
+      select voter_id, event_type, event_data
+      from event_log
+      where event_type IN (?, ?)
+      order by timestamp
+      `,
+      EventType.VoterCheckIn,
+      EventType.UndoVoterCheckIn
+    ) as Array<{
+      voter_id: string;
+      event_type: EventType;
+      event_data: string;
+    }>;
+    if (!rows) {
+      return voters;
+    }
+    for (const row of rows) {
+      switch (row.event_type) {
+        case EventType.VoterCheckIn: {
+          const voter = find(voters, (v) => v.voterId === row.voter_id);
+          if (!voter) {
+            continue;
+          }
+          voter.checkIn = safeParseJson(
+            row.event_data,
+            VoterCheckInSchema
+          ).unsafeUnwrap();
+          break;
+        }
+        case EventType.UndoVoterCheckIn: {
+          const voter = find(voters, (v) => v.voterId === row.voter_id);
+          if (!voter) {
+            continue;
+          }
+          voter.checkIn = undefined;
+          break;
+        }
+        default: {
+          throwIllegalValue(row.event_type);
+        }
+      }
+    }
+    return voters;
+  }
+
+  private getVoters(): Voter[] | undefined {
+    if (!data.voters) {
+      // Load the voters from the database if they are not in memory.
+      const rows = this.client.all(
+        `
+          select voter_data
+          from voters
+        `
+      ) as Array<{ voter_data: string }>;
+      if (!rows) {
+        return undefined;
+      }
+      data.voters = rows.map((row) =>
+        safeParseJson(row.voter_data, VoterSchema).unsafeUnwrap()
+      );
+    }
+    return this.applyEventsToVoters(data.voters);
+  }
+
+  saveEvents(pollbookEvents: PollbookEvent[]): boolean {
+    let isSuccess = true;
+    this.client.transaction(() => {
+      for (const pollbookEvent of pollbookEvents) {
+        isSuccess = isSuccess && this.saveEvent(pollbookEvent);
+      }
+    });
+    return isSuccess;
+  }
+
+  saveEvent(pollbookEvent: PollbookEvent): boolean {
+    try {
+      switch (pollbookEvent.type) {
+        case EventType.VoterCheckIn: {
+          const event = pollbookEvent as VoterCheckInEvent;
+          // Fail gracefully if the event was already saved.
```

Comment:
Given our discussion in Slack, I wouldn't expect this to ever happen. I wonder if we should fail loudly, since it might mean something weird happened

Reply:
Yeah will update. I was thinking it should always be safe if that did happen but you are right that at least for the next couple of weeks of dev and testing it would be good to know if it did